FileSystemAccessServer+TextEditor: Add new service for veiled file access

timmot opened the pull request:
### FileSystemAccessServer
* New service which provides file descriptors to veiled files, confirmed by user action.
* User can be prompted with a message box, if a program wants to access a currently veiled file via this service.
    - e.g. "Give RequestServer access to /home/anon/sample.html?"
* User can be prompted with an open/save file picker which will give them an fd if allowed by their user permissions.
* Any user approval is persisted until the application closes.
    - e.g. "Save as" a file in TextEditor and you will be able to "Save" it without being prompted again.
* Has read-write access to entire file system

### Future Goals
* Password-prompt when trying to access root files
* WebContent/RequestServer can use this to handle file:// URLs
* This service provides downloading as well because the sequence of selecting a download location matches well with the main functionality of this service.
    - User approval of downloads should not persisted, e.g. the browser shouldn't necessarily have access to write to a file simply because it initiated the download process.

timmot commented:
The amount of touch points via WebContent is quite high and it doesn't feel good to add or modify methods in OutOfProcessWebView, Page, and FrameLoader for downloading files.

The suggested DownloadServer sounds nicer and it wouldn't require edits to classes focused on loading and rendering data for display. Going to go ahead with this.

MaxWipfli commented on Userland/Services/FileSystemAccessServer/FileSystemAccessServer.ipc:
```diff
@@ -0,0 +1,10 @@
+endpoint FileSystemAccessServer
+{
```
Maybe the client would want to know if a request succeeded.
For `request_file_{read,write}` this could fail in the following ways:
* User declines permission.
* File system permissions deny permission.
* Any other way that the open syscall could fail.

For `prompt_file_{read,write}` this could also mean that the user cancelled the file picker dialog. For `download_file` there are also all the failures that RequestServer can have.

On another note, what is the `i32 a` for, and could it maybe have a more descriptive name?

timmot replied:
`i32 a` is there as a helpful friend for `IPC::File`.

When I tried to return only `IPC::File` it triggered `AK_MAKE_NONCOPYABLE` and caused the compilation to fail.

```
../.././AK/Optional.h: In instantiation of 'AK::Optional<T>::Optional(const AK::Optional<T>&) [with T = IPC::File]':
Userland/Services/FileSystemAccessServer/FileSystemAccessServerEndpoint.h:265:115:   required from 'AK::Optional<IPC::File> FileSystemAccessServerProxy<LocalEndpoint, PeerEndpoint>::request_file_read(AK::String) [with LocalEndpoint = FileSystemAccessClientEndpoint; PeerEndpoint = FileSystemAccessServerEndpoint]'
../../Userland/Applications/Browser/Tab.cpp:91:34:   required from here
../.././AK/Optional.h:55:13: error: use of deleted function 'IPC::File::File(const IPC::File&)'
   55 |             new (&m_storage) T(other.value());
      |             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from ../.././AK/RefCounted.h:12,
                 from ../.././AK/StringImpl.h:10,
                 from ../.././AK/String.h:14,
                 from ../.././AK/URL.h:10,
                 from ../../Userland/Applications/Browser/History.h:9,
                 from ../../Userland/Applications/Browser/Tab.h:9,
                 from ../../Userland/Applications/Browser/Tab.cpp:8:
../../Userland/Libraries/LibIPC/File.h:17:25: note: declared here
   17 |     AK_MAKE_NONCOPYABLE(File);
      |           
```

MaxWipfli commented on Userland/Services/FileSystemAccessServer/ClientConnection.cpp:
```diff
@@ -0,0 +1,125 @@
+
+Messages::FileSystemAccessServer::RequestFileReadResponse ClientConnection::request_file_read(String const& path)
+{
+    auto fd = open(path.characters(), O_RDONLY);
```
Maybe we should check that the user provided an absolute path with a simple `path.starts_with("/")`, since relative paths would be relative to the CWD of FileSystemAccessServer, which would only lead to bugs. I don't think we should `VERIFY`, but somehow respond with an error code or error message.

timmot replied:
The request_{read,write} methods are more developer facing than the prompts so I think it would make sense to reject any relative paths outright, or we could have an argument for the base path from the client?

MaxWipfli commented on Userland/Services/FileSystemAccessServer/ClientConnection.cpp:
```diff
@@ -0,0 +1,125 @@
+
+        if (fd < 0) {
+            dbgln(OSError(errno).string());
+            return { -1, Optional<IPC::File> {} };
```
Maybe you could change the IPC to return a `Result<IPC::File, int>`, which would allow you to e.g. use the positive numbers as the OS error codes, and the negative numbers for "own" errors, like:
* not an absolute path
* user terminated the dialog
* etc...

timmot replied:
Nice idea!

timmot replied:
Having more IPCCompiler issues when I try to use Result, I'm not sure I understand the error message entirely.


```cpp
// I've tried with and without the additional i32 error
request_file_read(String sender, String path) => (i32 error, Result<IPC::File,i32> fd_or_error)
```

```cpp
In file included from ../../Userland/Applications/Browser/Tab.cpp:41:
Userland/Services/FileSystemAccessServer/FileSystemAccessServerEndpoint.h: In static member function 'static AK::OwnPtr<Messages::FileSystemAccessServer::RequestFileReadResponse> Messages::FileSystemAccessServer::RequestFileReadResponse::decode(AK::InputMemoryStream&, int)':
Userland/Services/FileSystemAccessServer/FileSystemAccessServerEndpoint.h:69:46: error: could not convert '<brace-enclosed initializer list>()' from '<brace-enclosed initializer list>' to 'AK::Result<IPC::File, int>'
   69 |         Result<IPC::File,i32> fd_or_error = {};
      |                                              ^
      |                                              |
      |                                              <brace-enclosed initializer list>
In file included from Userland/Services/FileSystemAccessServer/FileSystemAccessClientEndpoint.h:13,
                 from ../../Userland/Applications/Browser/Tab.cpp:40:
../../Userland/Libraries/LibIPC/Decoder.h: In instantiation of 'bool IPC::decode(IPC::Decoder&, T&) [with T = AK::Result<IPC::File, int>]':
../../Userland/Libraries/LibIPC/Decoder.h:72:27:   required from 'bool IPC::Decoder::decode(T&) [with T = AK::Result<IPC::File, int>]'
Userland/Services/FileSystemAccessServer/FileSystemAccessServerEndpoint.h:70:28:   required from here
../../Userland/Libraries/LibIPC/Decoder.h:21:19: error: static assertion failed: Base IPC::decoder() instantiated
   21 |     static_assert(DependentFalse<T>, "Base IPC::decoder() instantiated");
      |                   ^~~~~~~~~~~~~~~~~
../../Userland/Libraries/LibIPC/Decoder.h:21:19: note: 'AK::Detail::DependentFalse<AK::Result<IPC::File, int> >' evaluates to false
In file included from Userland/Services/FileSystemAccessServer/FileSystemAccessClientEndpoint.h:15,
                 from ../../Userland/Applications/Browser/Tab.cpp:40:
../../Userland/Libraries/LibIPC/Encoder.h: In instantiation of 'bool IPC::encode(IPC::Encoder&, T&) [with T = const AK::Result<IPC::File, int>]':
../../Userland/Libraries/LibIPC/Encoder.h:84:20:   required from 'void IPC::Encoder::encode(const T&) [with T = AK::Result<IPC::File, int>]'
../../Userland/Libraries/LibIPC/Encoder.h:68:15:   required from 'IPC::Encoder& IPC::Encoder::operator<<(const T&) [with T = AK::Result<IPC::File, int>]'
Userland/Services/FileSystemAccessServer/FileSystemAccessServerEndpoint.h:89:19:   required from here
../../Userland/Libraries/LibIPC/Encoder.h:17:19: error: static assertion failed: Base IPC::encode() was instantiated
   17 |     static_assert(DependentFalse<T>, "Base IPC::encode() was instantiated");
      |                   ^~~~~~~~~~~~~~~~~
../../Userland/Libraries/LibIPC/Encoder.h:17:19: note: 'AK::Detail::DependentFalse<const AK::Result<IPC::File, int> >' evaluates to false
```

timmot replied:
I'm going to resolve this, but I'll keep your ideas about error numbers in mind.

MaxWipfli commented on Userland/Services/FileSystemAccessServer/ClientConnection.cpp:
```diff
@@ -0,0 +1,125 @@
+
+void ClientConnection::prompt_download_file(URL const& url, String const& filename, String const& ext)
+{
+    auto app = GUI::Application::construct(0, nullptr);
+    auto user_picked_file = GUI::FilePicker::get_save_filepath(nullptr, filename, ext, Core::StandardPaths::downloads_directory());
+    app->exec();
+    app->quit();
```
Do you need to always instantiate a `GUI::Application` here? I assume it's only about the event loop, so maybe a `Core::EventLoop` suffices.

timmot replied:
This was a little workaround to remove this error:
```
FileSystemAccessServer(35:35): USERSPACE(35) ASSERTION FAILED: !s_default_font_query.is_empty()
../../Userland/Libraries/LibGfx/FontDatabase.cpp:48
[#0 FileSystemAccessServer(35:35)]: Terminating FileSystemAccessServer(35) due to signal 6
```

It feels a little dirty to do it the current way, but I think the only alternative is to create a helper application?

timmot replied:
The GUI::Applications clash in prompt_download if you try and download two things. I might split the DownloadWidget to a completely seperate application.

timmot replied:
I replaced the main `Core::EventLoop` with a single `GUI::Application` like `NotificationServer` and got rid of all the other dodgy ones.

https://github.com/SerenityOS/serenity/pull/7964/commits/d5683b61ac493685d2e7423abf137858123c2553

MaxWipfli commented on Userland/Services/FileSystemAccessServer/ClientConnection.cpp:
```diff
@@ -95,31 +132,47 @@ Messages::FileSystemAccessServer::PromptFileWriteResponse ClientConnection::prom
-        auto file = Core::File::construct(user_picked_file.value());
-
-        file->open(Core::OpenMode::WriteOnly);
+    auto mainWindow = GUI::Window::construct();
```
`mainWindow` -> `main_window`

gunnarbeutner commented on Userland/Services/FileSystemAccessServer/FileSystemAccessServer.ipc:
```diff
@@ -0,0 +1,10 @@
+endpoint FileSystemAccessServer
+{
+    request_file_read(String path) => (i32 a, Optional<IPC::File> fd)
```
```suggestion
    request_file_read(String path) => (Optional<IPC::File> fd)
```

gunnarbeutner commented on Userland/Services/FileSystemAccessServer/ClientConnection.cpp:
```diff
@@ -0,0 +1,180 @@
+        // TODO: perform error checking
+        // TODO: if root file...
+
+        return { -1, IPC::File(fd, IPC::File::CloseAfterSending) };
```
```suggestion
        return Optional<IPC::File> { IPC::File(fd, IPC::File::CloseAfterSending) };
```

timmot replied:
Thank you Gunnar!

I was having trouble compiling IPC::File by itself, turns out I was using Optional incorrectly.

gunnarbeutner commented on Userland/Services/FileSystemAccessServer/ClientConnection.cpp:
```diff
@@ -0,0 +1,180 @@
+
+    if (result == GUI::MessageBox::ExecNo) {
+        // TODO: If a malicious program requests access over and over, do we eventually want to deny it?
+        return { -1, Optional<IPC::File> {} };
```
```suggestion
        return Optional<IPC::File> {};
```

gunnarbeutner commented on Userland/Services/FileSystemAccessServer/ClientConnection.cpp:
```diff
@@ -0,0 +1,180 @@
+        return { -1, Optional<IPC::File> {} };
+    }
+
+    return { -1, Optional<IPC::File> {} };
```
```suggestion
    return Optional<IPC::File> {};
```

MaxWipfli commented on Userland/Applications/TextEditor/main.cpp:
```diff
@@ -67,8 +112,33 @@ int main(int argc, char** argv)
-            return 1;
+        int fd;
+
+        if (!parsed_argument.filename().starts_with("/"sv)) {
```
What is the reason you are checking for relative paths here? AFAIK, you should be able to `open()` the path relatively, even though it needs to be absolute for unveiling.

timmot replied:
Just a mistake, I was thinking about absolute paths too much, fixed.

MaxWipfli commented on Userland/Applications/TextEditor/main.cpp:
```diff
@@ -67,8 +112,33 @@ int main(int argc, char** argv)
+
+            fd = open(file_path, O_RDONLY);
+
+            if (fd < 0)
```
For good measure, this should probably `perror("open")` here.

MaxWipfli commented on Userland/Applications/TextEditor/MainWidget.cpp:
```diff
@@ -272,34 +274,41 @@ MainWidget::MainWidget()
             return;
 
-        if (!m_editor->write_to_file(save_path.value())) {
+        if (!m_editor->write_to_file(response.fd().value().take_fd())) {
```
Just FYI, with `Optional`, you can use `*` and `->` to access the value at bit less verbose.
```cpp
response.fd()->take_fd()
```
It's not wrong how you did it, just as a heads-up.
If you just want the value of `response.fd()`, you could use `*response.fd()`.

timmot replied:
Ah very cool! I prefer this style, thanks.

MaxWipfli commented on Userland/Applications/TextEditor/MainWidget.cpp:
```diff
@@ -272,34 +274,41 @@ MainWidget::MainWidget()
-                window()->set_modified(false);
+            auto response = m_filesystem_client->request_file("TextEditor", m_path, FileSystemAccessServer::WriteCreateTruncate);
+
+            if (response.error() != 0) {
```
Does this mean that `error() = 0` indicates an error? This seems really weird because in general, `0` means success in most cases where you have numeric error codes.

timmot replied:
`error() = 0` is supposed to be a success as you expect, thanks for the catch.

MaxWipfli commented on Userland/Applications/TextEditor/MainWidget.cpp:
```diff
@@ -660,7 +670,15 @@ bool MainWidget::open_file(const String& path)
-    set_path(LexicalPath(path));
+    if (!path.starts_with("/"sv)) {
+        char* buf = getcwd(nullptr, 0);
+        LexicalPath lexical_path = LexicalPath::join(String(buf), path);
```
I think you don't need the explicit `String(buf)` here, as the `String(char const*)` is not marked `explicit`. You can just do:
```cpp
LexicalPath::join(buf, path);
```
Also, you could use `auto lexical_path` to avoid repeating yourself.

MaxWipfli commented on Userland/Applications/TextEditor/MainWidget.h:
```diff
@@ -16,14 +16,39 @@
+        : IPC::ServerConnection<FileSystemAccessClientEndpoint, FileSystemAccessServerEndpoint>(*this, "/tmp/portal/fileaccess")
+    {
+    }
+    virtual void dummy() override
```
I think the dummy methods should no longer be needed since #8221 has been merged.

MaxWipfli commented on Userland/Applications/TextEditor/main.cpp:
```diff
@@ -30,6 +31,50 @@ int main(int argc, char** argv)
+
+        if (!parsed_argument.filename().starts_with("/"sv)) {
+            char* buf = getcwd(nullptr, 0);
+            LexicalPath lexical_path = LexicalPath::join(String(buf), parsed_argument.filename());
```
Same as above with the unnecessary explicit `String(buf)` and `auto lexical_path`.

MaxWipfli commented on Userland/Libraries/LibGUI/TextEditor.cpp:
```diff
@@ -1179,14 +1179,8 @@ void TextEditor::timer_event(Core::TimerEvent&)
-        return false;
-    }
-
     ScopeGuard fd_guard = [fd] { close(fd); };
```
Since you no longer `open` the file descriptor in this message, it should probably also not be closed here, which is what the `ScopeGuard` does. It normally makes sense for the code that opens the file descriptor to also close it. Otherwise, it would make sense to declare in the method name that you will close the file descriptor (e.g. `write_to_file_and_close_fd`).

timmot replied:
Taking the latter path.

MaxWipfli commented on Userland/Services/FileSystemAccessServer/CMakeLists.txt:
```diff
@@ -0,0 +1,19 @@
+)
+
+serenity_bin(FileSystemAccessServer)
+target_link_libraries(FileSystemAccessServer LibCore LibIPC LibGUI LibWeb LibProtocol LibDesktop)
```
You should probably use the new `serenity_component` directive here. Also, the programs which now depend on the new Service should declare it as dependency in their `serenity_component` declaration.

MaxWipfli commented on Userland/Services/FileSystemAccessServer/ClientConnection.cpp:
```diff
@@ -0,0 +1,207 @@
+        if (requested_access & FileAccess::CreateOrRemove)
+            access |= O_CREAT;
+
+        auto fd = open(user_picked_file.value().characters(), access, 0644);
```
As I already explained above, you could use `user_picked_file->characters()` here (not required of course).

MaxWipfli commented on Userland/Services/FileSystemAccessServer/ClientConnection.cpp:
```diff
@@ -0,0 +1,207 @@
+    return { -1, Optional<IPC::File> {}, Optional<String> {} };
+}
+
+Messages::FileSystemAccessServer::PromptSaveFileResponse ClientConnection::prompt_save_file(String const& name, String const& ext, String const& path_to_view, int requested_access)
```
There is a lot of code duplication between `prompt_open_file` and `prompt_save_file`, maybe you could add a helper method for everything after the `FilePicker` dialog invocation, since that code seems to be identical.

timmot replied:
This should be less of a problem now that I've taken out the custom access flag logic, I think it will make sense to keep these separate until they bulk up again.

timmot replied:
Done.

MaxWipfli commented on Userland/Services/FileSystemAccessServer/ClientConnection.cpp:
```diff
@@ -0,0 +1,207 @@
+    auto user_picked_file = GUI::FilePicker::get_save_filepath(nullptr, filename, ext, Core::StandardPaths::downloads_directory());
+
+    if (user_picked_file.has_value()) {
+        auto window = GUI::Window::construct(main_window.ptr());
```
There's more code duplication here, maybe you could have a `open_download_window(URL const& url, String const& path)` method for constructing the window and starting the download.

It could also be a nice idea to add a `DownloadWindow : public GUI::Window` class which has a static method `static void open(URL const& url, String const& path)` or something like that, which does all that stuff for you.

MaxWipfli commented on Userland/Services/FileSystemAccessServer/DownloadWidget.cpp:
```diff
@@ -0,0 +1,154 @@
+
+namespace FileSystemAccessServer {
+
+DownloadWidget::DownloadWidget(const URL& url, String const& destination_path)
```
```suggestion
DownloadWidget::DownloadWidget(URL const& url, String const& destination_path)
```

MaxWipfli commented on Userland/Services/FileSystemAccessServer/DownloadWidget.cpp:
```diff
@@ -0,0 +1,154 @@
+    m_download = Web::ResourceLoader::the().protocol_client().start_request("GET", url.to_string());
+    VERIFY(m_download);
+    m_download->on_progress = [this](Optional<u32> total_size, u32 downloaded_size) {
+        did_progress(total_size.value(), downloaded_size);
```
You don't check `total_size.has_value()` here. If you are sure it always has a value, you could insert a `VERIFY(total_size.has_value())` to make it clear to others that this wasn't just overlooked.

timmot replied:
I'm very unsure in fact, I'm going surround `did_progress` with a check for `total_size.has_value()`

MaxWipfli commented on Userland/Services/FileSystemAccessServer/DownloadWidget.h:
```diff
@@ -0,0 +1,41 @@
+    virtual ~DownloadWidget() override;
+
+private:
+    explicit DownloadWidget(const URL&, String const&);
```
```suggestion
    DownloadWidget(URL const&, String const& destination_path);
```
It's not immediately obvious what the second parameters does, so let's give it a name. Also, `explicit` is only really necessary for single argument constructors.

MaxWipfli commented on Userland/Services/FileSystemAccessServer/FileSystemAccessClient.ipc:
```diff
@@ -0,0 +1,4 @@
+endpoint FileSystemAccessClient
+{
+    dummy() =|
```
This can go away now, see my other comment above.

MaxWipfli reviewed (commented):
Nice work you are doing here! I realised some of the comments about DownloadWidget aren't your code, but if you duplicate the code, let's fix those style errors.

timmot commented on Userland/Services/FileSystemAccessServer/DownloadWidget.cpp:
```diff
@@ -0,0 +1,161 @@
+    }
+
+    m_elapsed_timer.start();
+    m_download = Web::ResourceLoader::the().protocol_client().start_request("GET", url.to_string());
```
I've implemented this using its own `RequestClient` because it is nicer, as you say. I'm sure we can deal with any nastiness which turns up.

MaxWipfli commented on Userland/Applications/Browser/Tab.cpp:
```diff
@@ -57,12 +57,14 @@ URL url_from_user_input(const String& input)
-    window->set_resizable(false);
-    window->set_main_widget<DownloadWidget>(url);
-    window->show();
+    // URL basenames are not this consistent :-(
+    // TODO: More work identifying potential file names
+    auto filename_parts = url.basename().split('.');
+
+    if (filename_parts.size() == 2)
+        m_filesystem_client->async_prompt_download_file(url, filename_parts[0], filename_parts[1]);
+    else
+        m_filesystem_client->async_prompt_download_file(url, url.basename(), "txt");
```
You can either find the last `.` in the URL and split on that one manually, or use `LexicalPath`:
```suggestion
auto filename = LexicalPath(url.basename());
auto extension = filename.extension();
if (extension.is_empty())
    extension = "txt";
m_filesystem_client->async_prompt_download_file(url, filename.title(), extension);
```

MaxWipfli commented on Userland/Applications/Browser/Tab.cpp:
```diff
@@ -57,12 +57,14 @@ URL url_from_user_input(const String& input)
 
 void Tab::start_download(const URL& url)
```
If you touch files quite extensively, it would be nice if you would convert them to east const style (in a separate commit). :^)
```suggestion
void Tab::start_download(URL const& url)
```

MaxWipfli commented on Userland/Applications/Browser/Tab.cpp:
```diff
@@ -95,6 +97,7 @@ void Tab::view_dom_tree(const String& dom_tree)
 
 Tab::Tab(BrowserWindow& window, Type type)
     : m_type(type)
+    , m_filesystem_client(FileSystemAccessClient::construct())
```
Let's construct the `FileSystemAccessClient` lazily as soon as it's needed for the first time, since most browser tabs will never initiate a download in their lifetime.

MaxWipfli commented on Userland/Applications/Browser/main.cpp:
```diff
@@ -69,7 +69,17 @@ int main(int argc, char** argv)
     }
 
-    if (unveil("/home", "rwc") < 0) {
+    if (unveil(String::formatted("{}/BrowserContentFilters.txt", Core::StandardPaths::config_directory()).characters(), "rwc") < 0) {
```
```suggestion
    if (unveil(LexicalPath::join(Core::StandardPaths::config_directory(), "BrowserContentFilters.txt").characters(), "rwc") < 0) {
```
Same for the two `unveil` calls below. Also, you should probably cache the `Core::StandardPaths::config_directory()` in a variable to avoid threefold StringBuilder and String allocation inside that function call.

MaxWipfli commented on Userland/Applications/TextEditor/MainWidget.cpp:
```diff
@@ -40,13 +40,16 @@
 
 namespace TextEditor {
 
 MainWidget::MainWidget()
 {
     load_from_gml(text_editor_window_gml);
 
+    m_filesystem_client = FileSystemAccessClient::construct();
+
```
```suggestion
MainWidget::MainWidget()
    : m_filesystem_client(FileSystemAccessClient::construct())
{
    load_from_gml(text_editor_window_gml);
```
This allows you to change `m_filesystem_client` to `NonnullRefPtr<FileSystemAccessClient>`.

MaxWipfli commented on Userland/Applications/TextEditor/MainWidget.cpp:
```diff
@@ -272,28 +275,36 @@ MainWidget::MainWidget()
             return;
         }
         // FIXME: It would be cool if this would propagate from GUI::TextDocument somehow.
         window()->set_modified(false);
 
-        set_path(LexicalPath(save_path.value()));
-        dbgln("Wrote document to {}", save_path.value());
+        set_path(LexicalPath(response.chosen_file().value()));
+        dbgln("Wrote document to {}", response.chosen_file().value());
```
Just as an advance warning, this will become a merge conflict as soon as #8320 is merged.

MaxWipfli commented on Userland/Applications/TextEditor/MainWidget.cpp:
```diff
@@ -640,10 +651,11 @@ void MainWidget::update_title()
 }
 
-bool MainWidget::open_file(const String& path)
+bool MainWidget::open_file(int file_descriptor, const String& path)
```
Another east const reminder.

MaxWipfli commented on Userland/Applications/TextEditor/MainWidget.h:
```diff
@@ -53,6 +74,8 @@ class MainWidget final : public GUI::Widget {
 
     virtual void drop_event(GUI::DropEvent&) override;
 
+    RefPtr<FileSystemAccessClient> m_filesystem_client;
```
According to the comment on the constructor:
```suggestion
    NonnullRefPtr<FileSystemAccessClient> m_filesystem_client;
```

MaxWipfli commented on Userland/Applications/TextEditor/main.cpp:
```diff
@@ -30,6 +31,55 @@ int main(int argc, char** argv)
     parser.parse(argc, argv);
 
+    // Unveiling here so we can take advantage of the input file name
+    if (unveil("/home/anon/.config", "rw") < 0) {
```
```suggestion
    if (unveil(Core::standard_paths_config_directory().characters(), "rw") < 0) {
```

MaxWipfli commented on Userland/Libraries/LibGUI/TextEditor.cpp:
```diff
@@ -1229,6 +1223,17 @@ bool TextEditor::write_to_file(const String& path)
     return true;
 }
 
+bool TextEditor::write_to_file(const String& path)
```
```suggestion
bool TextEditor::write_to_file(String const& path)
```

MaxWipfli commented on Userland/Services/FileSystemAccessServer/ClientConnection.cpp:
```diff
@@ -0,0 +1,166 @@
+
+namespace FileSystemAccessServer {
+
+static HashMap<int, RefPtr<ClientConnection>> s_connections;
```
```suggestion
static HashMap<int, NonnullRefPtr<ClientConnection>> s_connections;
```
As a rule of thumb, you (almost) never want to store null pointers in a `HashMap`, `Vector` or similar.

MaxWipfli commented on Userland/Services/FileSystemAccessServer/ClientConnection.cpp:
```diff
@@ -0,0 +1,166 @@
+    VERIFY(path.starts_with("/"sv));
+
+    bool approved = false;
+    if (m_approved_files.get(path).has_value()) {
+        approved = (requested_access & m_approved_files.get(path).value()) == requested_access;
+    }
```
This will do two hash lookups, lets only do one:
```suggestion
    auto maybe_permissions = m_approved_files.get(path);
    if (maybe_permissions.has_value())
        approved = (requested_access & *maybe_permissions) == requested_access;
```

MaxWipfli commented on Userland/Services/FileSystemAccessServer/ClientConnection.cpp:
```diff
@@ -0,0 +1,166 @@
+
+    if (!approved) {
+        StringBuilder builder;
+        if (requested_access & O_RDONLY)
+            builder.append('r');
+
+        if (requested_access & O_WRONLY)
+            builder.append('w');
+
+        if (requested_access & O_CREAT)
+            builder.append('c');
```
Since you only support `O_RDONLY`, `O_WRONLY` and `O_CREAT`, you could maybe `VERIFY` that no other permissions were requested, with something like:
```cpp
VERIFY((requested_access & (O_RDONLY | O_WRONLY | O_CREAT)) == requested_access);

timmot replied:
I don't know if it is useful for a user to know if `Append`, `Truncate`, `MustBeNew`, or `KeepOnExec` is important for the user to note, but it is important for us to have access to these in the code.

MaxWipfli replied:
You're right, those should also be supported, but not shown to the user. So I guess you can disregard my suggestion. :^)

MaxWipfli commented on Userland/Services/FileSystemAccessServer/ClientConnection.cpp:
```diff
@@ -0,0 +1,166 @@
+
+        auto access_string = builder.to_string();
+
+        auto result = GUI::MessageBox::show(nullptr, String::formatted("Give {} \"{}\" access to \"{}\"?", sender, access_string, path), "File Permissions Requested", GUI::MessageBox::Type::Warning, GUI::MessageBox::InputType::YesNo);
```
This is a possible security issue, since the caller can decide himself what he is called. I dont' know IPC enough, maybe it's possible to somehow get the PID of the other process and then extract the process name from ProcFS?

MaxWipfli commented on Userland/Services/FileSystemAccessServer/ClientConnection.cpp:
```diff
@@ -0,0 +1,166 @@
+
+        approved = result == GUI::MessageBox::ExecYes;
+
+        if (approved) {
+            // Intentionally overwrite existing access
+            m_approved_files.set(path, requested_access);
+        }
```
What if one part of a process always requests "r" access, while another always requests "w" access. Maybe you could do this, to "merge" the two requested permissions:
```suggestion
        if (approved) {
            auto new_permissions = requested_access;
            if (maybe_permissions.has_value())
                new_permissions |= *maybe_permissions;
            m_approved_files.set(path, new_permissions);
        }
```

timmot replied:
I'm worried about `O_TRUNC`, `O_APPEND`, `O_EXCL`, and possibly more. Perhaps I should persist everything except for these flags.

timmot replied:
I'm not worried anymore, the open mode is already controlled by whatever requested access is.

MaxWipfli commented on Userland/Services/FileSystemAccessServer/ClientConnection.cpp:
```diff
@@ -0,0 +1,166 @@
+        auto fd = open(user_picked_file->characters(), requested_access, 0644);
+
+        if (fd < 0) {
+            dbgln(String::formatted("FileSystemAccessServer: Couldn't open {}, error {}", user_picked_file.value(), OSError(errno).string()));
```
Don't use `String::formatted` inside `dbgln`. Firstly, it's unnecessary since `dbgln` already supports formatting. Secondly, if the formatted String still contains curly braces, `dbgln` will interpret them, you you would need to put `dbgln("{}", String::formatted...`. But what you actually need to do it this:
```suggestion
            dbgln("FileSystemAccessServer: Couldn't open {}, error {}", user_picked_file.value(), OSError(errno).string());
```

MaxWipfli commented on Userland/Services/FileSystemAccessServer/ClientConnection.cpp:
```diff
@@ -0,0 +1,166 @@
+        auto fd = open(user_picked_file->characters(), requested_access, 0644);
+
+        if (fd < 0) {
+            dbgln(String::formatted("FileSystemAccessServer: Couldn't open {}, error {}", user_picked_file.value(), OSError(errno).string()));
```
Remove `String::formatted` as above.

MaxWipfli commented on Userland/Services/FileSystemAccessServer/ClientConnection.cpp:
```diff
@@ -0,0 +1,166 @@
+    return { -1, Optional<IPC::File> {}, Optional<String> {} };
+}
+
+void ClientConnection::request_download_file(String const& sender, URL const& url, String const& path)
```
What if the file already exists?

timmot replied:
I've changed the MessageBox text for when it's overwriting an existing file.

I can see `request_download_file` changing a lot when it starts to get used but I don't have a place to use it right now.
I think it might be best to delete it until a use case appears, how does that sound?

timmot replied:
Moved all download stuff to a separate branch, I'll start a separate PR for it later.

MaxWipfli commented on Userland/Services/FileSystemAccessServer/ClientConnection.cpp:
```diff
@@ -0,0 +1,166 @@
+
+        approved = result == GUI::MessageBox::ExecYes;
+
+        if (approved) {
+            m_approved_files.set(path, O_WRONLY);
+        }
```
Same as above, maybe use `|=` to unify the granted permissions.

timmot replied:
Done.

MaxWipfli commented on Userland/Services/FileSystemAccessServer/DownloadWidget.cpp:
```diff
@@ -7,44 +7,33 @@
     VERIFY(m_download);
     m_download->on_progress = [this](Optional<u32> total_size, u32 downloaded_size) {
-        did_progress(total_size.value(), downloaded_size);
+        if (total_size.has_value()) {
```
You don't need the curlies here. :^)

MaxWipfli commented on Userland/Services/FileSystemAccessServer/DownloadWidget.h:
```diff
@@ -23,13 +25,14 @@ class DownloadWidget final : public GUI::Widget {
 
     URL m_url;
     String m_destination_path;
+    RefPtr<Protocol::RequestClient> m_protocol_client;
```
This can be a `NonnullRefPtr`.

MaxWipfli commented on Userland/Services/FileSystemAccessServer/DownloadWindow.cpp:
```diff
@@ -0,0 +1,35 @@
+
+namespace FileSystemAccessServer {
+
+static HashMap<u32, RefPtr<DownloadWindow>> s_windows;
```
```suggestion
static HashMap<u32, NonnullRefPtr<DownloadWindow>> s_windows;
```

MaxWipfli commented on Userland/Services/FileSystemAccessServer/DownloadWindow.cpp:
```diff
@@ -0,0 +1,35 @@
+DownloadWindow::DownloadWindow(i32 client_id, URL const& url, String const& destination_path)
+{
+    m_id = client_id;
+    s_windows.set(m_id, this);
```
```suggestion
    s_windows.set(m_id, *this);
```
This is needed because a `NonnullRefPtr` cannot be constructed from a raw pointer, but it can be from a reference (since it's guaranteed non-null).

MaxWipfli commented on Userland/Services/FileSystemAccessServer/Forward.h:
```diff
@@ -0,0 +1,13 @@
+/*
```
I'm not sure this file is needed, since it's only included by `ClientConnection` itself.

MaxWipfli commented on Userland/Services/FileSystemAccessServer/main.cpp:
```diff
@@ -0,0 +1,29 @@
+
+int main(int, char**)
+{
+    if (pledge("stdio recvfd sendfd rpath cpath wpath unix thread", nullptr) < 0) {
```
Just a nitpick, but I'm not sure this needs `recvfd`, as it only sends FDs.

timmot replied:
I'm not certain why but I've tested without recvfd and it fails to start.

MaxWipfli reviewed (commented):
Here's another review. It's starts to take shape, which is great! :^)

timmot commented:
The changes I've rebased and force-pushed are:
* Creation of a `prompt_helper` function 
* Use of `Core::OpenMode` and `Core::File` instead of `int` and `open()`
* Remove of all download/browser stuff
* `m_approved_files` permissions are now aggregated
* Finding the client process using `peer_pid` and ProcFS

MaxWipfli commented on Userland/Services/FileSystemAccessServer/ClientConnection.cpp:
```diff
@@ -0,0 +1,132 @@
+        auto pid = this->socket().peer_pid();
+        auto exe_link = LexicalPath("/proc").append(String::number(pid)).append("exe").string();
+        auto exe_path = Core::File::real_path_for(exe_link);
+        auto exe_name = LexicalPath(exe_path).basename();
```
Since `basename()` returns a `StringView const&`, the memory pointed to by `exe_name` may already be freed. You should use `LexicalPath::basename(exe_path)`, which returns `String` instead of `StringView const&`.

MaxWipfli reviewed (changes requested):
After the one change left, this looks good to me. Thank you for working on this. :^)

MaxWipfli commented:
> After the one change left, this looks good to me. Thank you for working on this. :^)

I've found another problem: You currently store the full requested access in `m_approved_files`. This means if I request `O_WRONLY`, and it is granted, and I then re-request `O_WRONLY | O_TRUNC`, the user is prompted again for "w" access. I think you should only store the `O_RDONLY` and `O_WRONLY` flags in `m_approved_files`, and then check with
```cpp
approved = has_flag(*maybe_permissions, (requested_access & (Core::OpenMode::ReadOnly | Core::OpenMode::WriteOnly)))
```

MaxWipfli reviewed (approved):
LGTM. Let's get this merged, so we can start adding it to applications which can make use of it. :^)

awesomekling commented on Userland/Applications/TextEditor/MainWidget.cpp:
```diff
@@ -667,7 +667,15 @@ bool MainWidget::open_file(const String& path)
 
     m_editor->set_text(file->read_all());
 
-    set_path(path);
+    if (!path.starts_with("/"sv)) {
+        char* buf = getcwd(nullptr, 0);
+        auto lexical_path = LexicalPath::join(buf, path);
+        free(buf);
+
+        set_path(lexical_path.string());
+    } else {
+        set_path(path);
+    }
```
Can we use `Core::File::real_path_for()` to achieve all this instead? :)

timmot replied:
From my understanding we cannot because `realpath` requires unveiled access to the path.

I'm working around this by ensuring FileSystemAccessServer only returns absolute paths, and that a filename passed as an argument will be resolve before hand.

timmot replied:
Removed this change and have replaced it with `VERIFY(path.starts_with("/"sv));` at the top of `MainWidget::open_file`, along with the ideas above.

awesomekling commented on Userland/Applications/TextEditor/MainWidget.cpp:
```diff
@@ -647,10 +656,11 @@ void MainWidget::update_title()
 }
 
-bool MainWidget::open_file(const String& path)
+bool MainWidget::open_file(int file_descriptor, String const& path)
```
Please use `fd` for file descriptors.

timmot commented on Userland/Applications/TextEditor/main.cpp:
```diff
@@ -30,6 +32,55 @@ int main(int argc, char** argv)
+        FileArgument parsed_argument(file_to_edit);
+
+        if (!parsed_argument.filename().starts_with("/"sv)) {
+            char* buf = getcwd(nullptr, 0);
```
Use `Core::File::real_path_for()` here too.

awesomekling commented:
There's a lot of variations in the naming here:
- FileSystemAccessServer
- FSAccessServer
- filesystem_client (lack of separation between "file" and "system")
- /tmp/portal/fileaccess

Let's settle on a single name for this concept and use that throughout. :)

timmot commented:
* Changed the commit message to use FileSystemAccessSever.
* Using `m_file_system_access_client` as variable name
* Using `/tmp/portal/filesystemaccess`

timmot commented:
Since I keep rebasing:
* I changed the open prompt in TextEditor to open as ReadWrite because while it was a good functionality test to show a MessageBox on save (to confirm write-access), it was also irritating from a daily use point-of-view.
* I went ahead with and made some east-const style changes as recommended previously
* Removed comments in FileSystemAccessServer that were fairly useless

timmot commented:
* Using a better name `read_file_and_close` for reading from already open fd's.
* Replaced `open` use with `Core::File` use, and was able to remove all LibC style includes
